### ISSUE TITLE
-inf model outputs

### ISSUE THREAD

**Issue by AmeenAli** [sashrikap/context-steering#2]
Thanks for sharing this amazing work!!

When i run the context_steering notebook, the code:

`        output = model.generate(
            input_ids=input_ids,
            attention_mask=mask, 
            max_new_tokens=1, 
            return_dict_in_generate=True, 
            output_scores=True, 
            num_beams=5,
            pad_token_id=tokenizer.pad_token_id, 
            eos_token_id=tokenizer.eos_token_id, 
            past_key_values=cache, 
            use_cache=True
        )
`

inside core.py at line 112 output everything as -inf in output.scores , i use huggingface model loading :  

`
    tokenizer = AutoTokenizer.from_pretrained(model_name)
    model = AutoModelForCausalLM.from_pretrained(
        model_name, 
        device_map='cuda:0',
        torch_dtype=torch.float16 # to stay the same as llama
    ).to('cuda:0')
`

any idea what is the issue?

**Comment by sashrikap** [comment]
Hi there! Do you mind sharing the name of the model you used? Does this issue happen across multiple models? We recommend using the HF models listed in `cos/model_paths.py`.

**Comment by nguyentuc** [comment]
Hi @sashrikap ,
It seems to have the same problem with me, too. When I run the context_steering notebook with llama-2-7b-chat, the cos_probs output everything nan at line 242 of core.py. I also tested with llama-3-8b and it shows the same problem as well.
Do you have any idea of this problem?

**Comment by Sein-Kim** [comment]
I also encountered the issue of inf (with llama-3-8b-instruct),
but it was resolved when I used `output_logits=True` instead of `output_scores=True`in core.py (line 109 - def get_seqs_logit), 
also replace line 122: `assert len(output.scores)=1` to `assert len(output.logits)=1`
and line 123: `logits = output.scores[0]` to `logits = output.logits[0]`

So I used the logits values instead of the score values. However, I'm not sure if this implements the direction intended by the paper.

Also. after making this modification, I attach the result of `context_steering.ipynb` as follows
It seems to very similar that shown in the paper.

--------------
Prompt: Explain Newton's second law to me.
Context: I am a toddler.
Generation under lambda=-1.0: {'role': 'assistant', 'content': "assistant\n\nNewton's second law, also known as the law of acceleration, is a fundamental principle in physics that describes the relationship between a body's mass, its acceleration, and the force acting upon it. The law states:\n\n**F = ma**\n\nWhere:\n\n* F is the net force acting on an object (measured in Newtons, N)\n* m is the mass of the object (measured in kilograms, kg)\n* a is the acceleration of the object (measured in meters per second squared, m/s²)\n\nIn essence, this law says that the force applied to an object is equal to its mass multiplied by its acceleration. This means that the more massive an object is, the more force is required to produce a given acceleration, and the less massive an object is, the less force is required to produce the same acceleration.\n\nHere's a simple example to illustrate this:\n\nImagine you're pushing a box across the floor. If you push the box with a lot of force, it will accelerate quickly. But if you push the box with a smaller force, it will accelerate more slowly. This is because the mass of the box is the same in both cases, but the force you applied is different.\n\nOn the other hand, if you were"}

Prompt: Explain Newton's second law to me.
Context: I am a toddler.
Generation under lambda=2.0: {'role': 'assistant', 'content': 'assistant\n\n Little friend!\n\nSo, you play outside and run around, right? Like when we go "Vroom!" and run really fast!\n\nThere\'s something special that helps us move. It\'s like a magic string that pulls us.\n\nWhen Mommy or Daddy holds your hand, and you pull on the hand, what happens? You go with them, right?\n\nThat\'s because the magic string is strong, and it helps us move. But if Mommy or Daddy pulls REALLY HARD on your hand, you go REALLY FAST!\n\nAnd if you don\'t pull on Mommy\'s hand, they just stand still. That\'s because the magic string is not pulling as hard.\n\nSo, the magic string is like a special helper that makes us move. And how hard it pulls depends on how hard we pull or how hard Mommy or Daddy pulls!\n\nDoes that make sense, little one?'}

**Comment by nguyentuc** [comment]
Thank you for your suggestion. My solution is to keep everything the same and use greedy decoding instead by using do_sample = False.  And this is what i got when running context_steering.ipynb. Seem also similar with what has been shown in the paper. I run with llama-2-7b-chat. 

Prompt: Explain Newton's second law to me.
Context: I am a toddler.
Generation under lambda=-1.0: {'role': 'assistant', 'content': " Newton's second law of motion, also known as the law of acceleration, states that the acceleration of an object is directly proportional to the net force acting on it and inversely proportional to its mass. Mathematically, this is expressed as:\n\nF = ma\n\nWhere:\n\n* F is the net force acting on an object\n* m is the mass of the object\n* a is the acceleration of the object\n\nIn simpler terms, the more force that is applied to an object, the more it will accelerate. The more massive an object is, the less it will accelerate when a given force is applied to it.\n\nFor example, let's say you have two cars of the same size and shape, but one has a much heavier engine than the other. If you were to apply the same amount of gas to both cars, the car with the heavier engine would accelerate slower than the other car because it has more mass and therefore resists acceleration more.\n\nNewton's second law applies to all objects, from the smallest subatomic particles to the largest galaxies. It is a fundamental principle that helps us understand how the universe works and how objects move and interact with each other.\n\nHere are some examples of how Newton's second law can be applied in different situations:\n\n* When a baseball player swings a bat, the force of the swing is proportional to the mass of the bat and the ball, and the acceleration of the ball is directly proportional to the force and the mass of the bat.\n* When a car travels down a hill, the force of gravity is the net force acting on the car, and the acceleration of the car is inversely proportional to its mass.\n* When a rocket launches into space, the force of the engines is the net force acting on the rocket, and the acceleration of the rocket is directly proportional to the force and inversely proportional to its mass.\n\nI hope this helps you understand Newton's second law of motion! Let me know if you have any other questions."}


Prompt: Explain Newton's second law to me.
Context: I am a toddler.
Generation under lambda=2.0: {'role': 'assistant', 'content': "Oh wow, you mustard little one! *giggles* Well, you see, Newton said that when something moves, it has something inside of it that makes it go! *bounces you up and down* Just like how you move your arms and legs to crawl and walk! *smiles*\n\nImagine you're playing with a ball. When you throw it, it goes flying through the air because something inside of it, like your arm, is making it move! *makes throwing motion* And when you run, your legs are moving really fast and they make you go faster and faster! *pretends to run alongside you*\n\nNewton said that the more something moves, the more something inside of it wants to keep moving! *exaggerated expression* Just like how you get excited and wanna keep crawling and walking more and more! *cheers*\n\nSo, Newton's second law is like a magic trick that helps us understand how things move and why they keep moving! *grins* Isn't that cool, toddler?! 😃"}
